jobsdb/jobsdb.go

```diff
@@ -1128,21 +1131,31 @@ func (jd *HandleT) checkIfMigrateDS(ds dataSetT) (
 
 	recordsLeft = totalCount - delCount
 
-	// If records are newer than what is required. One example use case is
-	// gateway DB where records are kept to dedup
-
-	var minCreatedAt, maxCreatedAt time.Time
-	sqlStatement = fmt.Sprintf(`SELECT MIN(created_at), MAX(created_at) from %q`, ds.JobTable)
-	row = jd.dbHandle.QueryRow(sqlStatement)
-	err = row.Scan(&minCreatedAt, &maxCreatedAt)
-	jd.assertError(err)
+	if jd.MinDSRetentionPeriod > 0 {
+		var maxCreatedAt time.Time
+		sqlStatement = fmt.Sprintf(`SELECT MAX(created_at) from %q`, ds.JobTable)
+		row = jd.dbHandle.QueryRow(sqlStatement)
+		err = row.Scan(&maxCreatedAt)
+		jd.assertError(err)
 
-	if jd.MinDSRetentionPeriod > 0 && time.Since(maxCreatedAt) < jd.MinDSRetentionPeriod {
-		return false, false, recordsLeft
+		if time.Since(maxCreatedAt) < jd.MinDSRetentionPeriod {
+			return false, false, recordsLeft
+		}
 	}
 
-	if jd.MaxDSRetentionPeriod > 0 && time.Since(minCreatedAt) > jd.MaxDSRetentionPeriod {
-		return true, false, recordsLeft
+	if jd.MaxDSRetentionPeriod > 0 {
+		var terminalJobsExist sql.NullBool
+		sqlStatement = fmt.Sprintf(`SELECT EXISTS (
+									SELECT id
+										FROM %q
+										WHERE job_state IN ('%s') and now() - exec_time > interval '%d second' LIMIT 1)`,
```

Comment:
```suggestion
										WHERE job_state IN ('%s') and exec_time < $2)`,
```
Passing `time.Now().Add(jd.MaxDSRetentionPeriod)` as a second argument.
I also believe there is no need to use `limit 1` as the `exists` clause will take care of the necessary optimizations

Reply:
It should be `time.Now().Add(-1*jd.MaxDSRetentionPeriod)`. Done the change.

Reply:
you are correct, sorry for the misleading proposal